Add DataFile rewrite Action support for Iceberg

jerryshao opened the pull request:
This PR propose to add DataFile rewrite Action to Iceberg. The purpose of adding this functionality is to merge the small files into large ones to reduce the FileSystem overhead.

This PR uses `read.split.target-size` to control the read size of each task, and uses `write.target-file-size-bytes` to control the write size of each task. It also takes care of table partition to compact the files under each partition (without shuffling).  Besides user could specify the `Filter` expressions to only compact the files in some specific partitions.

rdblue commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesActionResult.java:
```diff
@@ -0,0 +1,50 @@
+    return EMPTY;
+  }
+
+  public List<DataFile> deleteDataFiles() {
```
Nit: should be `deletedDataFiles`.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,138 @@
+
+    } catch (Throwable t) {
+      originalThrowable = t;
+      try {
```
It looks like if the aborts succeed, then the task returns instead of throwing `t`. Is that correct behavior? I see that this is calling `context.markTaskFailed(originalThrowable)`. Is that done instead of throwing the exception?

jerryshao replied:
This is an issue, I will fix it.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,138 @@
+        }
+
+        // Wrap throwable in an Exception
+        throw new Exception(originalThrowable);
```
Why not throw the original throwable?

jerryshao replied:
This is mainly because the map function in `RDD` only supports `Exception` rather `Throwable`, so here I wrapped again to throw an Exception instead.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,138 @@
+        dataWriter.write(row);
+      }
+
+      dataReader.close();
```
Close should be called within a `finally` block, I think.

jerryshao replied:
This part of code is mostly referred from Spark DS related ones. 

I think we invoked `dataReader.close` in catch `Throwable` to close reader in any case, so there should be no necessary to use `finally` block.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,138 @@
+    int partitionId = context.partitionId();
+    long taskId = context.taskAttemptId();
+    long epochId = Long.parseLong(
+        Optional.ofNullable(context.getLocalProperty(MicroBatchExecution.BATCH_ID_KEY())).orElse("0"));
```
Why does this set the epoch ID? I think this is unnecessary and could just default to 0.

jerryshao replied:
Sure, will change to 0.

rdblue commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,186 @@
+    return table;
+  }
+
+  public RewriteDataFilesAction specId(int specId) {
```
This sets the partition spec that will be used when writing new data files, but that isn't obvious from the configuration method name. How about `outputSpecId` or something more descriptive? I think this should also be documented for callers.

rdblue commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,186 @@
+  private List<CloseableIterable<FileScanTask>> groupTasksByPartition(CloseableIterable<FileScanTask> tasks) {
+    Map<StructLike, List<FileScanTask>> tasksGroupedByPartition = Maps.newHashMap();
+
+    tasks.forEach(task -> {
```
This method iterates through `tasks` and builds the grouped result in memory. The result doesn't need to be a `CloseableIterable` because it is in memory and there shouldn't be any resources to close.

This also doesn't close `tasks` or the iterator used (which is what closing `tasks` would do). Instead, it combines each value list into a `CloseableIterable` with the original tasks. I think the intent was to close `tasks` when the result of this method is consumed. But `tasks` is never really closed because the `collect` doesn't handle a `CloseableIterable` any differently than an `Iterable`. In addition, by combining each partition list with `tasks`, closing each one would close `tasks` again, which isn't a good idea.

Let's refactor this by passing in `tasks.iterator()`. That will be a `CloseableIterator` that should be closed after it is consumed. Like this:

```
try {
  tasksIter.forEachRemaining(task -> {
    ...
  });
} finally {
  tasksIter.close()
}
```

rdblue commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,186 @@
+    int lookbak = PropertyUtil.propertyAsInt(
+        table.properties(),
+        TableProperties.SPLIT_LOOKBACK,
+        TableProperties.SPLIT_LOOKBACK_DEFAULT);
```
While I agree that using the table values as defaults is a good idea, I think this should also expose configuration methods for combined size and lookback.

This is important because lookback(1) is used to prevent files from being reordered. Because only one bin is open at a time, either a file fits or it doesn't so the order of files is not changed. That's important when compacting a partition written with a global ORDER BY, which will distribute the sort column data across files. Iceberg uses that distribution to prune files when users filter by those fields.

rdblue commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,186 @@
+      RewriteFiles rewriteFiles = table.newRewrite();
+      rewriteFiles.rewriteFiles(Sets.newHashSet(deletedDataFiles), Sets.newHashSet(addedDataFiles));
+      commit(rewriteFiles);
+    } catch (Exception e) {
```
We typically use the `boolean threw` pattern instead of using a broad catch like this. The purpose is to still try to clean up even if the problem wasn't an `Exception` (like `AssertionError`):

```
boolean threw = true;
try {
  RewriteFiles = table.newRewrite().rewriteFiles(...).commit()
  threw = false;
} finally {
  Tasks.foreach(...)
      .run(fileIO::deleteFile);
}
```

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,138 @@
+  private final Writer.WriterFactory writerFactory;
+  private final boolean caseSensitive;
+
+  public RowDataRewriter(Table table, JavaSparkContext sparkContext, PartitionSpec spec, boolean caseSensitive,
```
The purpose of this class doesn't look well-defined. The Spark context is passed in, but so are already broadcasted variables. The main function here looks like it handles calling the DSv2 components and is serializable. If that's the case, then I think it would be cleaner to refactor the Spark context out and just pass in an RDD of CombinedScanTask.

rdblue commented on spark/src/test/java/org/apache/iceberg/actions/TestRewriteDataFilesAction.java:
```diff
@@ -0,0 +1,300 @@
+  private String tableLocation = null;
+
+  public TestRewriteDataFilesAction() {
+  }
```
You can omit empty constructors.

rdblue commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,186 @@
+        .filter(filter)
+        .planFiles();
+    List<DataFile> currentDataFiles =
+        Lists.newArrayList(CloseableIterable.transform(fileScanTasks, FileScanTask::file));
```
This is going to iterate through `fileScanTasks` before `groupTasksByPartition` also iterates through `fileScanTasks`, so it will cause job planning to happen twice. Instead, I think that this check should be moved after `groupTasksByPartition`, which would return a `Map`. If the map has no entries, then there is no work to do.

This should also be more aggressive. If a partition has just one file to compact, then that partition should be removed because there is no work to do.

jerryshao replied:
I see, will change to that way.

rdblue commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,186 @@
+    tasks.forEach(task -> {
+      List<FileScanTask> taskList = tasksGroupedByPartition.getOrDefault(task.file().partition(), Lists.newArrayList());
+      taskList.add(task);
+      tasksGroupedByPartition.put(task.file().partition(), taskList);
```
This should use `StructLikeWrapper` for the key instead of `StructLike` to handle `CharSequence` values correctly.

Jiayi-Liao commented on spark/src/main/java/org/apache/iceberg/spark/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,138 @@
+            partitionId, taskId, context.attemptNumber(), context.stageId(), context.stageAttemptNumber());
+        dataReader.close();
+        dataWriter.abort();
+        LOG.error("Aborted commit for partition {} (task {}, attempt {}, stage {}.{})",
```
Same as Line 118?

jerryshao replied:
No, it is not the same. "Aborting" -> "Aborted".

rdblue commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,268 @@
+   * @param rewriteDataFileSize size of rewrite data file
+   * @return this for method chaining
+   */
+  public RewriteDataFilesAction rewriteDataFileSizeInBytes(long rewriteDataFileSize) {
```
I think that `rewriteDataFile` is redundant because this is a rewrite data files action. How about `targetSizeInBytes` instead?

rdblue commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,268 @@
+  }
+
+  /**
+   * Specify the split lookback
```
I think it would help to give a bit more context here. This is the number of "bins" considered when trying to pack the next file split into a task. Increasing this usually makes tasks a bit more even by considering more ways to pack file regions into a single task, but with extra planning cost.

This should also mention that the bin packing can reorder the incoming file regions, so to preserve order for lower/upper bounds in file metadata, the caller can use a lookback of 1.

rdblue commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,268 @@
+
+  /**
+   * Pass a row Expression to filter DataFiles to be rewritten. This is typically used when only rewriting DatFiles
+   * under some partitions.
```
I think it is important to mention that all files that may contain data matching the filter may be rewritten.

When planning, the filter is used to match files inclusively: if a file may contain a matching row, the file matches. When matching file stats, the same inclusive approach is used. So if a user has a bucked table, `bucket(id, 16)` and asks to rewrite where `id = X`, then this will match any data file in `id_bucket = bucket(X, 16)`.

rdblue commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,268 @@
+      threw = false;
+
+    } catch (Throwable t) {
+      LOG.error("Failed to rewrite DataFiles", t);
```
This needs to re-throw the exception, not just swallow it.

rdblue commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,268 @@
+  public RewriteDataFilesActionResult execute() {
+    CloseableIterable<FileScanTask> fileScanTasks = table.newScan()
+        .filter(filter)
+        .planFiles();
```
Could this use `ManifestGroup` instead of a scan? The scan will emit a scan event, but this isn't really a table scan.

Also, this should be wrapped in a try-with-resources block so that the `CloseableIterable` is closed.

jerryshao replied:
AFAIK, we already called `close` in `groupTasksByPartition`, do we need to call wrap the code here in `try-with-resources` again?

jerryshao replied:
Besides, `ManifestGroup` is package private, so it cannot be used here.

rdblue replied:
Yes, we need to close the `CloseableIterable`. That will close all iterators that are still open. The close in `groupTasksByPartition` closes the `CloseableIterator` that was passed in, not the group of iterators created from the iterable. Ideally, there shouldn't be other open iterators, but closing the iterable helps ensure that we catch any that are accidentally missed.

rdblue commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,268 @@
+    Map<StructLikeWrapper, List<FileScanTask>> groupedTasks = groupTasksByPartition(fileScanTasks.iterator());
+    // Nothing to rewrite if the table is empty.
+    if (groupedTasks.isEmpty()) {
+      return RewriteDataFilesActionResult.empty();
```
Since this is checked below, it's not necessary to have 2. Not a big deal to keep it though.

rdblue commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,268 @@
+        .collect(Collectors.toList());
+    List<CombinedScanTask> combinedScanTasks = groupedCombinedTasks.stream()
+        .flatMap(Streams::stream)
+        .collect(Collectors.toList());
```
Why not combine this with the previous statement?

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,125 @@
+    long taskId = context.taskAttemptId();
+    DataWriter<InternalRow> dataWriter = writerFactory.createDataWriter(partitionId, taskId, 0);
+
+    Throwable originalThrowable = null;
```
Why not `catch (Throwable originalThrowable)` instead? I don't see the value of having this outside of the catch block.

jerryshao replied:
I see, will change it.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,125 @@
+        }
+      }
+
+      throw new RuntimeException(originalThrowable);
```
Instead of wrapping by default, this should check if the throwable is an `Exception` and throw it, then wrap if it isn't an exception.

rdblue replied:
I think this should check whether `originalThrowable` is an `Exception` and throw it if so. That way we don't needlessly wrap it. Only throwables that aren't exceptions should be wrapped.

rdblue commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,260 @@
+      try {
+        tasksIter.close();
+      } catch (IOException ioe) {
+        LOG.warn("Faile to close task iterator", ioe);
```
Nit: typo `Faile` -> `Failed`.

rdblue commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,260 @@
+   * Specify the number of "bins" considered when trying to pack the next file split into a task.
+   * Increasing this usually makes tasks a bit more even by considering more ways to pack file regions into a single
+   * task with extra planning cost.
+   *
```
Javadoc requires `<p>` on empty lines if you want a paragraph break.

rdblue commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,260 @@
+    try {
+      tasksIter.forEachRemaining(task -> {
+        StructLikeWrapper structLike = StructLikeWrapper.wrap(task.file().partition());
+        List<FileScanTask> taskList = tasksGroupedByPartition.getOrDefault(structLike, Lists.newArrayList());
```
This creates a new `ArrayList` for every element in the iterator to pass in as a default value. It would be better to use a `ListMultimap` instead so that you pass a supplier to create the lists:

```java
  ListMultimap<...> groups = Multimaps.newListMultimap(Maps.newHashMap(), Lists::newArrayList);
  ...
  groups.put(wrapper, task);
```

rdblue commented:
I think this is almost ready. I think it just needs to close the tasks iterable and use a multimap. I think @aokolnychyi wanted to have a look as well.

rdblue commented:
Anton pointed out a problem with this: if the tasks include a residual filter, it would be used to filter row groups in files, potentially losing data. He's working on a fix to ignore residuals in the table scan that should be ready tomorrow for you to include in this.

jerryshao commented:
OK, got it.

aokolnychyi commented:
The PR to ignore residuals is [here](https://github.com/apache/iceberg/pull/1094).

I'll go through this one tomorrow morning if you don't mind guys.

aokolnychyi commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,275 @@
+      threw = false;
+
+    } finally {
+      if (threw) {
```
Can we simplify this block by using `catch (Exception e)` and rethrowing the original exception? Otherwise, we won't know what happened.

rdblue replied:
We typically use this pattern with a `threw` boolean to always clean up even if the throwable isn't an `Exception`. It's more important to do this when the finally block needs to execute for all cases, though. I think it would be probably be fine to catch `Exception` here like it was originally -- sorry for the bad review comment.

aokolnychyi commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,275 @@
+    CloseableIterable<FileScanTask> fileScanTasks = null;
+    try {
+      fileScanTasks = table.newScan()
+          .caseSensitive(caseSensitive)
```
This is the place where we need to call `ignoreResiduals()` to make sure we read complete files.

aokolnychyi commented on spark/src/test/java/org/apache/iceberg/actions/TestRewriteDataFilesAction.java:
```diff
@@ -0,0 +1,297 @@
+
+import static org.apache.iceberg.types.Types.NestedField.optional;
+
+public class TestRewriteDataFilesAction {
```
I would advise adding a test where a file we rewrite has multiple row groups and our filter expression matches only one of them. It is not an easy test but should be worth it.

This can be done by setting `TableProperties.PARQUET_ROW_GROUP_SIZE_BYTES` to a small value like 100, inserting a relatively large number of rows sorted by a column, and using a filter on that column that matches only last 1-2 records.

jerryshao commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,275 @@
+      return RewriteDataFilesActionResult.empty();
+    }
+
+    long openFileCost = PropertyUtil.propertyAsLong(
```
Yes, it seems meaningful to set `openFileCost` to zero, but will it increase the overhead to pack many small files into one task?

jerryshao replied:
@rdblue what's your opinion on this?

rdblue replied:
I think there is still value for the open file cost. If there are a lot of small files, we don't want to pack them all into a single task for this action, just like we don't want to scan them all in a single task when reading. I think the only thing we would need to do is to allow the user to override this setting. We could do that in this one or a follow-up.

jerryshao replied:
I will add a configuration method for user to override.

jerryshao commented:
@rdblue @aokolnychyi would you please help to review again, thanks. I've addressed all the comments except one about `openFileCost` setting to zero or not.

rdblue commented on spark/src/test/java/org/apache/iceberg/actions/TestRewriteDataFilesAction.java:
```diff
@@ -0,0 +1,344 @@
+      records.add(new ThreeColumnRecord(i, String.valueOf(i), String.valueOf(i % 4)));
+    }
+    Dataset<Row> df = spark.createDataFrame(records, ThreeColumnRecord.class);
+    writeDF(df);
```
How do we know there will be multiple files for c3=0?

jerryshao replied:
Yes, because the Spark master we set is `local[2]`, so we will only have 2 task slots, which will only have 2 files.

rdblue commented on spark/src/test/java/org/apache/iceberg/actions/TestRewriteDataFilesAction.java:
```diff
@@ -0,0 +1,344 @@
+      Assert.assertEquals("Residuals must be ignored", Expressions.alwaysTrue(), task.residual());
+    }
+    List<DataFile> dataFiles = Lists.newArrayList(CloseableIterable.transform(tasks, FileScanTask::file));
+    Assert.assertEquals("Should have 2 data files before rewrite", 2, dataFiles.size());
```
I don't think it is necessary to have this assertion or to run the scan planning above. This tests that `ignoreResiduals` works properly, but we can assume that for this test.

aokolnychyi replied:
I think the best way to verify is by checking that the test fails if we remove `ignoreResiduals()` from the action.

rdblue commented:
+1 from me, but I'll wait to merge until @aokolnychyi has a chance to review.

aokolnychyi commented on spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,281 @@
+        table.properties(),
+        TableProperties.WRITE_TARGET_FILE_SIZE_BYTES,
+        TableProperties.WRITE_TARGET_FILE_SIZE_BYTES_DEFAULT);
+    this.targetSizeInBytes = Math.min(splitSize, targetFileSize);
```
Hm, if my split size is 128 MB and target file size is 512 MB, what is the size of files that we will create?

rdblue replied:
Hm. Split size is how much data per task while reading, and write target size is how much data to put in a file when writing. The behavior of the write target size is more of an upper bound -- files could be smaller and Iceberg has no ability to coalesce to larger sizes. I think this should ignore the target file size because that's enforced in the writer regardless. Only the split size should matter.

aokolnychyi replied:
I'd consider packing files into bins that are around `targetFileSize` as defined by the table property and ignoring the threshold during writes by setting `target-file-size-bytes` to the max long value so that we avoid closing files near the end of bins. That way, we will get files that are close to the target file size.

rdblue replied:
After thinking about this a bit more, I think I wasn't correct that target write size is a maximum. It's currently used that way in Spark, but the property is "target" and that implies that it is the size files _should_ be, ideally. So Anton is right and that's the size of files we should attempt to create. This also makes sense because it is a `write` property, not a `read` property.

I think we should remove the min with split size and just use the table's target size as the default.

aokolnychyi replied:
Let's also set `target-file-size-bytes` to the max long value while writing so that we avoid closing files near the end of bins. For example, if we ended up with 520 MB of data in a task, it is better to write it as one file instead of 512MB and 8 MB.

aokolnychyi replied:
I am also OK addressing this in a follow-up as the PR itself is large.

jerryshao replied:
@aokolnychyi @rdblue , let me rephrase the purpose you mentioned above:

1. Using the `targetFileSize` to do the split, and also the written file size.
2. Setting `write.target-file-size-bytes` to `Long.MAX_VALUE` to make sure there's no separated files for one written task.

Am I understanding right? 

Besides, what is the default value of `targetFileSize`, should it be explicitly defined by user, or use `SPLIT_SIZE` as default?

rdblue replied:
I think the conclusion was that the planning in this feature should use `write.target-file-size-bytes` for the default `targetFileSize` when planning the splits because we should use the target write size (instead of the target read/split size for most reads) for our task size.

Next, we also need to override `write.target-file-size-bytes` for the write itself because we expect the write to already produce files close to that size, and leaving it at the normal target is likely to produce tiny files when a task's output is just over the limit.

jerryshao replied:
Thanks @rdblue ,

The default value of `write.target-file-size-bytes` is `Long.MAX_VALUE`. If users don't override this configuration explicitly, and use this to do planning ( bin-packing), will this lead to the problem where all the `FileScanTask`s combined into one?

aokolnychyi commented:
I am +1 as well. Only one [question](https://github.com/apache/iceberg/pull/1083#discussion_r438890925) I have.

rdblue commented:
I agree with @aokolnychyi about how we should change the configuration for this. As he also suggested, I'm going to merge this PR and we can fix the config in a follow-up since this is ready to go and quite a large change.

Thanks for your work on this, @jerryshao!

jerryshao commented:
Thanks @rdblue @aokolnychyi for your review, I will create a follow-up PR to address this left thing.
